### PR TITLE
feat: support and enforce number and boolean transform functions

### PR DESCRIPTION
**Pull request by quentinderoubaix**

### CODE REVIEW

File: angular/headless/src/lib/coercion.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ *  ```typescript
+ *  @Input({ transform: auBooleanAttribute }) status: boolean | undefined;
+ *  ```
+ * @param value Value to be transformed.
+ */
+export function auBooleanAttribute(value: unknown): boolean | undefined {
+	if (value === undefined) {
+		return undefined;
+	}
+	return typeof value === 'boolean' || value === undefined ? value : value != null && value !== 'false';
```
Comment (divdavem) [inline angular/headless/src/lib/coercion.ts:15]: Why do you again compare with undefined? It is already done line 12.

Comment (quentinderoubaix) [reply angular/headless/src/lib/coercion.ts:15]: Indeed, thanks !

**Comment by divdavem** [comment]
You should also change `reportMissingInputProp` to generate the correct code in case the type is boolean/number.

And you should add the corresponding test cases:
```ts
		{
			code: codeTemplate('', 'myProp: boolean;'),
			errors: [{messageId: 'missingProp', data: {type: 'input', name: 'myProp'}}],
			output: codeTemplate("\n\t@Input({alias: 'auMyProp', transform: auBooleanAttribute}) myProp: boolean | undefined;\n", 'myProp: boolean;'),
		},
		{
			code: codeTemplate('', 'myProp: number;'),
			errors: [{messageId: 'missingProp', data: {type: 'input', name: 'myProp'}}],
			output: codeTemplate("\n\t@Input({alias: 'auMyProp', transform: auNumberAttribute}) myProp: number | undefined;\n", 'myProp: number;'),
		},
```

File: eslint-plugin/src/angular-check-props.ts [hunk 2]
```diff
@@ -320,7 +346,42 @@ export const angularCheckPropsRule = ESLintUtils.RuleCreator.withoutDocs({
 									}
 									const decoratorArguments = (getDecorator(classMember, 'Input')!.expression as TSESTree.CallExpression).arguments;
 									const inputValidAlias = validAlias(name);
-									if (!decoratorArguments.length || decoratorArguments[0].type !== 'Literal' || decoratorArguments[0].value !== inputValidAlias) {
+									const typeAsString = typeToString(inputInfo.type, classMember, context);
+									if (typeAsString === 'boolean | undefined' || typeAsString === 'number | undefined') {
+										const isBoolean = typeAsString === 'boolean | undefined';
```
Comment (divdavem) [inline eslint-plugin/src/angular-check-props.ts:351]: I would move this up and use `isBoolean` also in the if statement.

File: eslint-plugin/src/angular-check-props.ts [hunk 3]
```diff
@@ -170,16 +170,42 @@ const reportInvalidOutputPropType = (
...
+		node,
+		messageId: type,
+		data: {
+			name,
+			metadata,
+		},
+		*fix(fixer) {
+			const decorator = getDecorator(node, 'Input')!;
+			yield fixer.remove(decorator);
+			yield fixer.insertTextBefore(node, `@Input({alias: '${alias}', transform: ${transform}})`);
```
Comment (divdavem) [inline eslint-plugin/src/angular-check-props.ts:194]:
```suggestion
			yield fixer.insertTextBefore(node, `@Input(${metadata})`);
```

**Comment by quentinderoubaix** [comment]
thx for the reivew @divdavem ! I have made some changes following your recommendations, when you have time to check ;)

File: angular/headless/src/lib/coercion.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ */
+export function auNumberAttribute(value: unknown): number | undefined {
+	// parseFloat(value) handles most of the cases we're interested in (it treats null, empty string,
+	// and other non-number values as NaN, where Number just uses 0) but it considers the string
+	// '123hello' to be a valid number. Therefore we also check if Number(value) is NaN.
+	if (value === undefined) {
+		return undefined;
+	}
+	const isNumberValue = !isNaN(parseFloat(value as any)) && !isNaN(Number(value));
+	return isNumberValue ? Number(value) : NaN;
```
Comment (divdavem) [inline angular/headless/src/lib/coercion.ts:36]: Is there any reason why we do not do simply `return +value`?

Comment (divdavem) [reply angular/headless/src/lib/coercion.ts:36]: Maybe because of null and empty string...

Comment (divdavem) [reply angular/headless/src/lib/coercion.ts:36]: So maybe we could simply do:
```ts
return (value === null || value === "") ? NaN : +value;
```

Comment (quentinderoubaix) [reply angular/headless/src/lib/coercion.ts:36]: I have to confess I put my brain on the side and copied [angular's implementation](https://github.com/angular/angular/blob/main/packages/core/src/util/coercion.ts)...

I also hesitated to write instead:
```typescript
export function auNumberAttribute(value: unknown): number | undefined {
	if (value === undefined) {
		return undefined;
	}
	return numberAttribute(value);
}
```
What do you think ?

**Comment by divdavem** [comment]
> thx for the reivew @divdavem ! I have made some changes following your recommendations, when you have time to check ;)

Thank you for taking my comments into account!
Apart from the logic to coerce to number, I am fine with the PR.

File: angular/headless/src/lib/coercion.ts (hunk 4 above)
Comment (divdavem) [reply angular/headless/src/lib/coercion.ts:36]: I prefer that indeed.

Comment (quentinderoubaix) [reply angular/headless/src/lib/coercion.ts]: ok for me !

**Pull request merged**